### PR TITLE
Generate topographic zone weights

### PR DESCRIPTION
**Pull request by gavinevans**
This PR aims to generate weights for each topographic zone using the supplied orography. This is another step towards #283. 

Further work is required to use the weights when collapsing the topographic zone coordinate, in order to generate a single field following neighbourhood processing each topographic zone separately.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/generate_ancillaries/generate_ancillary.py [hunk 1]
```diff
@@ -36,7 +36,8 @@
 from cf_units import Unit
```
Comment (fionaRust) [inline lib/improver/generate_ancillaries/generate_ancillary.py]: Pylint is complaining about the import order. I don't claim to understand it's logic but it wan't iris after glob I think. I know you haven't introduced this but we could tidy it up anyway.

File: lib/improver/generate_ancillaries/generate_topographic_zone_weights.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module for generating the weights for topographic zones."""
+
+from cf_units import Unit
+import numpy as np
+import iris
+from iris.exceptions import InvalidCubeError
+import warnings
```
Comment (fionaRust) [inline lib/improver/generate_ancillaries/generate_topographic_zone_weights.py]: Again pylint is complaining about the import order. Sorry

File: lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+
+    def test_not_equal_to_max_band_number(self):
+        """Test that the results are as expected when the band number is not
+        equal to the max band number."""
+        expected_weights = np.array([[[1.0, 1.0],
+                                      [0.75, 0.5]],
+                                     [[0.0, 0.0],
+                                      [0.25, 0.5]]])
+        topographic_zone_weights = np.array([[[1.0, 1.0],
+                                            [0.75, 0.5]],
```
Comment (fionaRust) [inline lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py]: This indentation has gone a bit weird

File: lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+    def test_all_above_midpoint(self):
+        """Test that the results are as expected when all of the points
+        are above the midpoint."""
+        expected_weights = np.array([[[0.75, 0.7],
+                                      [0.6, 0.55]],
+                                     [[0.25, 0.3],
+                                      [0.4, 0.45]]])
+        topographic_zone_weights = np.array([[[0.75, 0.7],
+                                            [0.6, 0.55]],
```
Comment (fionaRust) [inline lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py]: So has this indentation

File: lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+    def test_not_equal_to_zeroth_band_number(self):
+        """Test that the results are as expected when the band number is not
+        equal to the zeroth band number."""
+        expected_weights = np.array([[[0.25, 0.0],
+                                      [0.0, 0.0]],
+                                     [[0.75, 1.0],
+                                      [1.0, 1.0]]])
+        topographic_zone_weights = np.array([[[0.0, 0.0],
+                                            [0.0, 0.0]],
```
Comment (fionaRust) [inline lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py]: So has this indentation

File: lib/improver/generate_ancillaries/generate_ancillary.py (hunk 1 above)
Comment (fionaRust) [reply lib/improver/generate_ancillaries/generate_ancillary.py]: Now fixed :)

File: lib/improver/generate_ancillaries/generate_topographic_zone_weights.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+            # Calculate the contribution to the weights from the adjacent
+            # upper band.
+            topographic_zone_weights.data = (
+                self.add_weight_to_upper_adjacent_band(
+                    topographic_zone_weights.data, orography_band,
+                    midpoints[band_number], band_number,
+                    len(bands)-1))
+
+        # Metadata updates
+        topographic_zone_weights.rename("Topographic_zone_weights")
```
Comment (fionaRust) [inline lib/improver/generate_ancillaries/generate_topographic_zone_weights.py]: Just noticed looking at the acceptance test data that we don't normally have capital letters in cube names.

File: lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py [hunk 7]
```diff
@@ -369,6 +369,42 @@ def test_data_no_mask(self):
         self.assertArrayAlmostEqual(
             result.data, expected_weights_data, decimal=2)
 
+    def test_data_no_mask_three_bands(self):
+        """Test that the result data is as expected, when none of the points
+        are masked and there are three bands defined."""
+        orography_data = np.array([[[[10., 40., 45],
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py]: I've added a decimal point for consistency.

**Pull request merged**